payjoin/src/send/mod.rs

```diff
@@ -163,30 +158,65 @@ mod error;
 
 type InternalResult<T> = Result<T, InternalValidationError>;
 
-/// Builder for sender-side payjoin parameters
-///
-/// These parameters define how client wants to handle Payjoin.
-pub struct Configuration {
+pub struct RequestBuilder<'a> {
+    psbt: Psbt,
+    uri: PjUri<'a>,
     disable_output_substitution: bool,
     fee_contribution: Option<(bitcoin::Amount, Option<usize>)>,
     clamp_fee_contribution: bool,
     min_fee_rate: FeeRate,
 }
 
-impl Configuration {
+impl<'a> RequestBuilder<'a> {
+    /// Prepare an HTTP request and request context to process the response
+    ///
+    /// An HTTP client will own the Request data while Context sticks around so
+    /// a `(Request, Context)` tuple is returned from `RequestBuilder::build()`
+    /// to keep them separated.
+    pub fn from_psbt_and_uri(
+        psbt: Psbt,
+        uri: crate::Uri<'a, NetworkChecked>,
+    ) -> Result<Self, CreateRequestError> {
+        let uri = uri
+            .check_pj_supported()
+            .map_err(|_| InternalCreateRequestError::UriDoesNotSupportPayjoin)?;
+        Ok(Self {
+            psbt,
+            uri,
+            // Sender's optional parameters
+            disable_output_substitution: false,
+            fee_contribution: None,
+            clamp_fee_contribution: false,
+            min_fee_rate: FeeRate::ZERO,
+        })
+    }
+
+    /// Disable output substitution even if the receiver didn't.
+    ///
+    /// This forbids receiver switching output or decreasing amount.
+    /// It is generally **not** recommended to set this as it may prevent the receiver from
+    /// doing advanced operations such as opening LN channels and it also guarantees the
+    /// receiver will **not** reward the sender with a discount.
+    pub fn always_disable_output_substitution(mut self, disable: bool) -> Self {
+        self.disable_output_substitution = disable;
+        self
+    }
+
     // Calculate the recommended fee contribution for an Original PSBT.
     //
     // BIP 78 recommends contributing `originalPSBTFeeRate * vsize(sender_input_type)`.
     // The minfeerate parameter is set if the contribution is available in change.
     //
     // This method fails if no recommendation can be made or if the PSBT is malformed.
-    pub fn recommended(
-        psbt: &Psbt,
-        payout_scripts: impl IntoIterator<Item = ScriptBuf>,
+    pub fn build_recommended(
```

Comment:
we can implement the `build_x` in a nicer way using enums and then have two build functions when one is pub with enum mapping and one private

```rust
enum BuildMethods {
   recommended: {
   ...
   }
   additionalfees {
   ...
   }
}
 ```

Reply:
I'm not sure I follow. How does this simplify the interface from a downstream perspective? I'm not sure I'm familiar with this pattern.

Reply:
ah am seeing now that `non_incen..` and `addtional_fee` are called from `recommended`, so that wont work. 

though about this 
```rust
    9 enum BuildConfiguration {
    8     Recommended(FeeRate),
    7     AdditionalFee {
    6        max_fee_contribution: bitcoin::Amount,
    5        change_index: Option<usize>,
    4        min_fee_rate: FeeRate,
    3        clamp_fee_contribution: bool,
    2     },
    1     NonIncentivizing,
  117 }
```
and then we can `match` to specific items in enum and call `build()` in each different case with different inputs. anyway, u can ignore this :D

Reply:
glad you're thinking through the design. thanks for the review